File: rag/advanced_rag/harness/sufficiency_ladder.py
"""Decision ladder for sufficiency — LLM AutoRater as primary judge, agent
confidence as a risk gate (Sufficient Context paper, arXiv 2411.06037).

This replaces the old "code cross-check + dual-signal weighted fusion + 5-way
verdict" pipeline. Because the operator cannot train/calibrate fusion weights
(the paper's logistic regression), we use a **monotonic threshold ladder**: the
LLM AutoRater decides *whether the evidence is sufficient to infer a plausible
answer*; the agent's self-assessed confidence only modulates *how the verdict is
presented* (full answer vs. caveated answer vs. reconcile). The only continuous
degree of freedom is conservativeness thresholds (``c_high`` / ``c_low`` /
``llm_floor``), which are product policy knobs, not learned parameters.

Semantics (paper §3 / §5.1):
  - Sufficiency = "a plausible answer A' can be inferred from the context",
    independent of whether that answer is correct, and without a ground truth.
  - Selective generation = confidence does not re-score the LLM's sufficiency
    verdict; it only decides whether to answer fully, caveat, or investigate.

Actions:
  - ``ANSWER``             full answer (sufficient + high confidence)
  - ``ANSWER_WITH_CAVEAT`` answer with an explicit caveat
  - ``GAP``                follow-up search on the concrete missing pieces
  - ``RECONCILE``          re-investigate a low-confidence / contradictory point
  - ``UNANSWERABLE``       cannot answer (or fall back to a direct LLM)
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

# Action constants (shared with callers / orchestrators).
ANSWER = "ANSWER"
ANSWER_WITH_CAVEAT = "ANSWER_WITH_CAVEAT"
GAP = "GAP"
RECONCILE = "RECONCILE"
UNANSWERABLE = "UNANSWERABLE"


@dataclass
class LadderOutput:
    """Result of the decision ladder."""

    action: str
    should_continue: bool
    caveat: str = ""
    missing: list[str] = field(default_factory=list)


def aggregate_agent_confidence(
    agent_results: list[Any],
    hard_violations: set[str] | dict[str, list[str]] | None = None,
) -> float:
    """Mean self-confidence over the trusted (self-verified, no hard veto) claims.

    No weights, no normalization — we only pick the trustworthy subset and take
    the mean. Hard vetoes (required-entity gaps, grounded-fact violations,
    numeric conflicts) suppress a claim's confidence entirely, so a confidence
    signal can never mask a missing-evidence gap (selective generation).
    """
    violations: set[str] = set()
    if hard_violations:
        if isinstance(hard_violations, dict):
            violations = set(hard_violations.keys())
        else:
            violations = set(hard_violations)

    trusted = [r for r in agent_results if getattr(r, "is_verified", False) and getattr(r, "claim_id", "") not in violations]
    if not trusted:
        return 0.0
    total = 0.0
    for r in trusted:
        try:
            total += float(getattr(r, "confidence", 0.0) or 0.0)
        except (TypeError, ValueError):
            continue
    return total / len(trusted)


def sufficiency_ladder(
    *,
    auto_sufficient: bool,
    auto_confidence: float,
    missing: list[str],
    contradictions: list[str],
    agent_confidence: float,
    c_high: float,
    c_low: float,
    llm_floor: float,
    allows_reconcile: bool,
    cycle: int,
    max_cycles: int,
    hard_violations: set[str] | dict[str, list[str]] | None = None,
) -> LadderOutput:
    """Evaluate the decision ladder and return the action.

    Parameters
    ----------
    auto_sufficient : bool
        LLM AutoRater's verdict (a plausible answer can be inferred).
    auto_confidence : float
        AutoRater's own confidence in its sufficiency judgment.
    missing : list[str]
        Concrete gaps when AutoRater says insufficient.
    contradictions : list[str]
        Evidence-internal contradictions (even if sufficient, caveat).
    agent_confidence : float
        Aggregated agent self-confidence (0..1).
    c_high / c_low : float
        Agent-confidence gates for full vs. caveated answer.
    llm_floor : float
        If AutoRater confidence < this, re-investigate regardless of verdict.
    allows_reconcile : bool
        Whether the mode can force a re-investigation (medium=False).
    cycle / max_cycles : int
        0-based current cycle and the mode's budget.
    hard_violations : set | dict
        Claim IDs (or {id: gaps}) with a proven evidence gap (required entity
        missing / grounded absent / numeric conflict). Any non-empty value
        forces a caveated answer even if AutoRater says sufficient.
    """
    violations: set[str] = set()
    if hard_violations:
        violations = set(hard_violations.keys()) if isinstance(hard_violations, dict) else set(hard_violations)

    # 1. Hard veto floor: code-proven evidence gap beats the LLM's "good enough".
    if violations:
        return LadderOutput(
            action=ANSWER_WITH_CAVEAT,
            should_continue=False,
            caveat=f"hard evidence gap in claim(s): {sorted(violations)[:6]}",
            missing=missing,
        )

    # 2. AutoRater says insufficient.
    if not auto_sufficient:
        if missing:
            return LadderOutput(action=GAP, should_continue=True, missing=missing)
        return LadderOutput(action=UNANSWERABLE, should_continue=False, missing=missing)

    # 3. AutoRater is not confident in its own sufficiency call.
    if auto_confidence < llm_floor:
        if allows_reconcile and cycle < max_cycles - 1:
            return LadderOutput(
                action=RECONCILE,
                should_continue=True,
                caveat="AutoRater itself is unsure; re-investigating",
            )
        return LadderOutput(
            action=ANSWER_WITH_CAVEAT,
            should_continue=False,
            caveat="AutoRater sufficiency judgment is low-confidence",
        )

    # 4. Sufficient + confident: agent confidence sets the presentation.
    caveat = ""
    should_continue = False
    if agent_confidence >= c_high:
        action = ANSWER
    elif agent_confidence >= c_low:
        action = ANSWER_WITH_CAVEAT
        caveat = "evidence partially supports the answer"
    elif allows_reconcile and cycle < max_cycles - 1:
        action = RECONCILE
        should_continue = True
    else:
        action = ANSWER_WITH_CAVEAT
        caveat = "evidence partially supports the answer"

    if contradictions:
        # Never silently pick one side of a contradiction; surface it.
        caveat = "evidence contains conflicting figures"
        action = ANSWER_WITH_CAVEAT
        should_continue = False

    return LadderOutput(action=action, should_continue=should_continue, caveat=caveat, missing=missing)

File: internal/agent/harness/sufficiency_ladder.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package harness

import (
	"sort"
	"strings"
)

// Action constants (shared with orchestrators), mirroring Python
// sufficiency_ladder.py. The ladder decides *presentation*: full answer vs.
// caveated answer vs. re-investigate vs. unanswerable.
const (
	ActionAnswer           = "ANSWER"
	ActionAnswerWithCaveat = "ANSWER_WITH_CAVEAT"
	ActionGap              = "GAP"
	ActionReconcile        = "RECONCILE"
	ActionUnanswerable     = "UNANSWERABLE"
)

// LadderInput is the decision-ladder input set, mirroring the keyword arguments
// of Python sufficiency_ladder(). AutoSufficient/AutoConfidence are the LLM
// AutoRater's judgment; Missing/Contradictions are its concrete gaps; and
// AgentConfidence is the aggregated agent self-confidence (risk gate).
type LadderInput struct {
	AutoSufficient  bool
	AutoConfidence  float64
	Missing         []string
	Contradictions  []string
	AgentConfidence float64
	CHigh           float64
	CLow            float64
	LLMFloor        float64
	AllowsReconcile bool
	Cycle           int
	MaxCycles       int
	HardViolations  map[string][]string // claimID -> gaps (or empty slices as markers)
}

// LadderOutput is the decision-ladder result.
type LadderOutput struct {
	Action         string
	ShouldContinue bool
	Caveat         string
	Missing        []string
}

// AggregateAgentConfidence mirrors Python aggregate_agent_confidence: mean
// self-confidence over the trusted (self-verified, non-violating) claims.
// HardViolations keys are claim IDs that must be excluded regardless of their
// self-reported verification.
func AggregateAgentConfidence(results []AgentResult, hardViolations map[string][]string) float64 {
	violated := map[string]bool{}
	for id := range hardViolations {
		violated[id] = true
	}
	total := 0.0
	count := 0
	for _, r := range results {
		if !r.IsVerified || violated[r.ClaimID] {
			continue
		}
		total += r.Confidence
		count++
	}
	if count == 0 {
		return 0.0
	}
	return total / float64(count)
}

// SufficiencyLadder evaluates the monotonic decision ladder and returns the
// action. It is a pure function (no LLM dependency), mirroring Python
// sufficiency_ladder() decision-by-decision.
func SufficiencyLadder(in LadderInput) LadderOutput {
	violations := map[string][]string{}
	for id, gaps := range in.HardViolations {
		violations[id] = gaps
	}

	// 1. Hard veto floor: code-proven evidence gap beats the LLM's "good enough".
	if len(violations) > 0 {
		ids := make([]string, 0, len(violations))
		for id := range violations {
			ids = append(ids, id)
		}
		sort.Strings(ids)
		first := ids
		if len(first) > 6 {
			first = first[:6]
		}
		return LadderOutput{
			Action:         ActionAnswerWithCaveat,
			ShouldContinue: false,
			Caveat:         "hard evidence gap in claim(s): " + strings.Join(first, ", "),
			Missing:        in.Missing,
		}
	}

	// 2. AutoRater says insufficient.
	if !in.AutoSufficient {
		if len(in.Missing) > 0 {
			return LadderOutput{Action: ActionGap, ShouldContinue: true, Missing: in.Missing}
		}
		return LadderOutput{Action: ActionUnanswerable, ShouldContinue: false, Missing: in.Missing}
	}

	// 3. AutoRater is not confident in its own sufficiency call.
	if in.AutoConfidence < in.LLMFloor {
		if in.AllowsReconcile && in.Cycle < in.MaxCycles-1 {
			return LadderOutput{
				Action:         ActionReconcile,
				ShouldContinue: true,
				Caveat:         "AutoRater itself is unsure; re-investigating",
			}
		}
		return LadderOutput{
			Action:         ActionAnswerWithCaveat,
			ShouldContinue: false,
			Caveat:         "AutoRater sufficiency judgment is low-confidence",
		}
	}

	// 4. Sufficient + confident: agent confidence sets the presentation.
	caveat := ""
	shouldContinue := false
	action := ActionAnswer
	switch {
	case in.AgentConfidence >= in.CHigh:
		action = ActionAnswer
	case in.AgentConfidence >= in.CLow:
		action = ActionAnswerWithCaveat
		caveat = "evidence partially supports the answer"
	case in.AllowsReconcile && in.Cycle < in.MaxCycles-1:
		action = ActionReconcile
		shouldContinue = true
	default:
		action = ActionAnswerWithCaveat
		caveat = "evidence partially supports the answer"
	}

	if len(in.Contradictions) > 0 {
		// Never silently pick one side of a contradiction; surface it.
		caveat = "evidence contains conflicting figures"
		action = ActionAnswerWithCaveat
		shouldContinue = false
	}

	return LadderOutput{Action: action, ShouldContinue: shouldContinue, Caveat: caveat, Missing: in.Missing}
}
